Add Support for Renaming module

scottming opened the pull request:
This PR is primarily copied from #374, but compared to the previous implementation, we have made two changes:

1. We constrained the scope of renaming, meaning that renaming a module is only allowed at its definition. This simplification reduces a lot of computations.
2. Based on the first point, our preparation no longer returns just a single local module, but instead, the entire module. This means that module renaming has become more powerful. We can not only rename a single local module but also simplify a module. For example, renaming `TopLevel.Parent.Child` to `TopLevel.Child`, or renaming some middle parts, like `TopLevel.Parent.Child` to `TopLevel.Renamed.Child`.

![CleanShot 2024-03-17 at 13 55 15@2x](https://github.com/lexical-lsp/lexical/assets/12830256/e44c3151-6ab6-45cf-97ae-c15c61cfe9a4)


I personally really like this change, especially the second point, which makes module renaming much more **practical**.

scottming committed:
Introduced rename functionality across modules

Enhanced the application by integrating rename functionality, increasing its editing capabilities. This includes supporting the preparation and execution of rename operations in various modules such as the lexical AST module, protocol request and response modules, remote control API, and server provider handling. Key additions entail aliasing for easier reference, handling new rename-related requests and responses, and implementing server state updates to acknowledge rename capabilities. The changes underscore our dedication to improving the tool's utility and ensuring a more dynamic and responsive user experience in code manipulation scenarios.

Filter out the `:struct` type references

Add some handler tests for `rename`

Add doc for `local_module_name`

Move `rename` from `code_intelligence` to `code_mod`

Move the prepare logic to a individual module

This PR is primarily copied from #374, but compared to the previous implementation, we have made two changes:

1. We constrained the scope of renaming, meaning that renaming a module is only allowed at its definition. This simplification reduces a lot of computations.
2. Based on the first point, our preparation no longer returns just a single local module, but instead, the entire module. This means that module renaming has become more powerful. We can not only rename a single local module but also simplify a module. For example, renaming TopLevel.Parent.Child to TopLevel.Child, or renaming some middle parts, like TopLevel.Parent.Child to TopLevel.Renamed.

I personally really like this change, especially the second point, which makes module renaming much more practical.

Surround the whole module when renaming happens

Remove logic related to the rename function.

Apply some code review suggestions

Fix a bug when expanding the module

Fix a merge error

Apply a format module usage suggestion

Update apps/remote_control/lib/lexical/remote_control/code_mod/rename.ex

Co-authored-by: Steve Cohen <scohen@scohen.org>

Apply the `defdelegate` suggestion

Fix a bug when the descendant containing the old suffix

scohen commented on apps/remote_control/lib/lexical/remote_control/code_intelligence/rename/module.ex:
```diff
@@ -0,0 +1,164 @@
+  end
+
+  defp to_edits_by_uri(results, new_name) do
+    Enum.group_by(
```
nitpick, the group_by should not be in this function, it should be in the caller. It's  more flexible to deal with an individual element than a group of things. The caller would be
```elixir
|> Enum.group_by(& &Document.Path.ensure_uri(&1.path), &result_to_edits/1))
```

scohen reviewed (changes requested):
I think we need to talk more about the ergonomics of this PR and whether and how we do file renames. They can be tricky.

scottming commented on apps/remote_control/lib/lexical/remote_control/code_intelligence/rename/module.ex:
```diff
@@ -0,0 +1,164 @@
+defmodule Lexical.RemoteControl.CodeIntelligence.Rename.Module do
```
I think you make a fair point, and I need to give more thought to the ergonomics aspect.

Additionally, a part that this PR is missing is that it currently only allows for renaming the module directly **under the cursor** (between two points, or between a point and a space). If we could support renaming the entire module completely, it might be more effective, though this could also be more challenging.

scottming commented on apps/remote_control/lib/lexical/remote_control/code_intelligence/rename/module.ex:
```diff
@@ -0,0 +1,164 @@
+    # Users won't always want to rename descendants of a module.
+    # For instance, when there are no more submodules after the cursor.
+    # like: `defmodule TopLevel.Mo|dule do`
+    # in most cases, users only want to rename the module itself.
```
I think you make a fair point, and I need to give more thought to the ergonomics aspect.

Additionally, a part that this PR is missing is that it currently only allows for renaming the module directly **under the cursor** (between two points, or between a point and a space, etc). If we could support renaming the entire module completely, it might be more effective, though this could also be more challenging.

scottming replied:
has been fixed, the new method is much simpler.

scottming commented on apps/remote_control/lib/lexical/remote_control/api.ex:
```diff
@@ -47,6 +47,22 @@ defmodule Lexical.RemoteControl.Api do
+        %Position{} = position
+      ) do
+    RemoteControl.call(project, CodeIntelligence.Rename.Module, :prepare, [analysis, position])
+  end
```
Changed to `CodeIntelligence.Rename, :prepare`

scottming commented on apps/remote_control/lib/lexical/remote_control/code_intelligence/rename/module.ex:
```diff
@@ -0,0 +1,164 @@
+
+  @spec rename(Analysis.t(), Position.t(), String.t()) ::
+          {:ok, %{Lexical.uri() => [Edit.t()]}} | {:error, term()}
+  def rename(%Analysis{} = analysis, %Position{} = position, new_name) do
```
I'm considering solving this problem in the next PR. I'm not quite sure about the support for `RenameFile` in `documentChanges` across various editors. Additionally, can we also accomplish this through a code action approach?

scohen replied:
why would you not be sure about the support across editors? It's been there since LSP 3.13, and is part of a `WorkspaceEdit` message. If you're unsure about rename, you should be equally unsure about using `WorkspaceEdit`s. The spec also says

FYI, both LSP clients for emacs support workspace edits and document renames. I'm also certain vscode supports these as well. 

We can also investigate helix and zed rather easily, and you can check out nvim.

scottming replied:
Resovled in #651

scottming commented on apps/remote_control/lib/lexical/remote_control/code_mod/rename.ex:
```diff
@@ -0,0 +1,28 @@
+  def rename(%Analysis{} = analysis, %Position{} = position, new_name) do
+    case Prepare.resolve(analysis, position) do
+      {:ok, {renamable, entity}, range} ->
+        rename_module = @renamable_mapping[renamable]
```
> shouldn't you see if the thing can be renamed before you prepare it? 

`Prepare.resolve` does exactly that. When it returns `{:ok, tuple, range}`, it means that this is an entity we support for renaming.

Therefore, I think using `@renaming_mapping[renamable]` here is fine, as it will certainly succeed unless we wrote the keys in `renaming_mapping` incorrectly, but that would be easily discovered by testing.

scottming commented on apps/remote_control/lib/lexical/remote_control/code_mod/rename/prepare.ex:
```diff
@@ -0,0 +1,89 @@
+  def prepare(%Analysis{} = analysis, %Position{} = position) do
+    case resolve(analysis, position) do
+      {:ok, {:module, module}, range} ->
+        {:ok, inspect(module), range}
```
Nope, that's why we use inspect here.

scohen replied:
You should be using `Lexical.Formats.module(module)`

scottming commented on apps/remote_control/lib/lexical/remote_control/code_mod/rename/prepare.ex:
```diff
@@ -0,0 +1,89 @@
+      end)
+
+    if is_nil(result) do
+      case Entity.resolve(analysis, position) do
+        {:ok, other, _} ->
+          Logger.info("Unsupported entity for renaming: #{inspect(other)}")
+          {:error, :unsupported_entity}
+
+        {:error, reason} ->
+          {:error, reason}
+      end
```
Yes, it's just for the *error message*, because I found that our `Entity.resolve` still has some missing cases. I'm inclined to keep this logic for a while.

scohen replied:
Doesn't `Entity.resolve` already have a log message if it can't resolve something? Why have two?

scottming replied:
Sorry, perhaps I didn't explain clearly enough.

The reason we need to call it twice here is that the *specific rename module is only concerned with the resolve results of that module*. For instance, the `module` module is only concerned with `module and struct` cases, while the `callable` module only cares about the `:call` situation.

So, when we use this code here, it only returns the entities we supported or `nil`. 

https://github.com/lexical-lsp/lexical/blob/8dc3c2f850ec1b266938b7c72cbcd9789aef3891/apps/remote_control/lib/lexical/remote_control/code_mod/rename/prepare.ex#L65-L72

The purpose of the second `resolve` is to identify which entities we do not support, something we can't see within the specific modules.

FYI: another callable rename branch:

https://github.com/lexical-lsp/lexical/blob/ee4dc9f909d308e13e128974539792ef2f8c489d/apps/remote_control/lib/lexical/remote_control/code_mod/rename/callable.ex#L15-L23

scohen commented:
I think i've found the issue that caused the hard crash, I'll write up a PR for that soon. 

I've tried this a couple times, renaming `Lexical.Ast` to `Lexical.AST` which I think is a pretty standard operation, and it fails in both emacs and vscode. Once you get that working, tag for a re-review.

Of course, you'll have to wait until after my other fix.

scottming commented:
> I've tried this a couple times, renaming Lexical.Ast to Lexical.AST which I think is a pretty standard operation, and it fails in both emacs and vscode. Once you get that working, tag for a re-review.

@scohen I tried merging 663, and then everything seems to be working well now.

scohen commented:
This still failed in vscode. I got a bunch of the following errors, which caused a supervisor to shut down:

```
09:06:22.383 [error] Failed to handle Elixir.LXical.Protocol.Notifications.DidChange, {{:error, :not_open}, %LXical.Server.State{configuration: %LXical.Server.Configuration{project: %LXical.Project{root_uri: "file:///Users/steve/Projects/lexical", mix_exs_uri: "file:///Users/steve/Projects/lexical/mix.exs", mix_project?: true, mix_env: nil, mix_target: nil, env_variables: %{}, project_module: nil, entropy: 49917, project_config: []}, support: %LXical.Server.Configuration.Support{code_action_dynamic_registration: true, hierarchical_symbols: true, snippet: true, deprecated: true, tags: #Protocol.Types.Completion.ClientCapabilities.TagSupport<[value_set: [:deprecated]]>, signature_help: #Protocol.Types.SignatureHelp.ClientCapabilities<[context_support: true, dynamic_registration: true, signature_information: #Protocol.Types.SignatureHelp.ClientCapabilities.SignatureInformation<[active_parameter_support: true, documentation_format: [:markdown, :plain_text], parameter_information: #Protocol.Types.SignatureHelp.ClientCapabilities.ParameterInformation<[label_offset_support: true]>]>]>, work_done_progress: true}, client_name: "Visual Studio Code", additional_watched_extensions: nil, dialyzer_enabled?: false}, initialized?: true, shutdown_received?: false, in_flight_requests: %{}}}
09:06:22.383 [error] Could not handle message LXical.Protocol.Notifications.DidChange :ok
09:06:22.395 [error] Failed to handle Elixir.LXical.Protocol.Notifications.DidChange, {{:error, :not_open}, %LXical.Server.State{configuration: %LXical.Server.Configuration{project: %LXical.Project{root_uri: "file:///Users/steve/Projects/lexical", mix_exs_uri: "file:///Users/steve/Projects/lexical/mix.exs", mix_project?: true, mix_env: nil, mix_target: nil, env_variables: %{}, project_module: nil, entropy: 49917, project_config: []}, support: %LXical.Server.Configuration.Support{code_action_dynamic_registration: true, hierarchical_symbols: true, snippet: true, deprecated: true, tags: #Protocol.Types.Completion.ClientCapabilities.TagSupport<[value_set: [:deprecated]]>, signature_help: #Protocol.Types.SignatureHelp.ClientCapabilities<[context_support: true, dynamic_registration: true, signature_information: #Protocol.Types.SignatureHelp.ClientCapabilities.SignatureInformation<[active_parameter_support: true, documentation_format: [:markdown, :plain_text], parameter_information: #Protocol.Types.SignatureHelp.ClientCapabilities.ParameterInformation<[label_offset_support: true]>]>]>, work_done_progress: true}, client_name: "Visual Studio Code", additional_watched_extensions: nil, dialyzer_enabled?: false}, initialized?: true, shutdown_received?: false, in_flight_requests: %{}}}
09:06:22.395 [error] Could not handle message LXical.Protocol.Notifications.DidChange :ok
09:06:22.417 [error] Failed to handle Elixir.LXical.Protocol.Notifications.DidChange, {{:error, :not_open}, %LXical.Server.State{configuration: %LXical.Server.Configuration{project: %LXical.Project{root_uri: "file:///Users/steve/Projects/lexical", mix_exs_uri: "file:///Users/steve/Projects/lexical/mix.exs", mix_project?: true, mix_env: nil, mix_target: nil, env_variables: %{}, project_module: nil, entropy: 49917, project_config: []}, support: %LXical.Server.Configuration.Support{code_action_dynamic_registration: true, hierarchical_symbols: true, snippet: true, deprecated: true, tags: #Protocol.Types.Completion.ClientCapabilities.TagSupport<[value_set: [:deprecated]]>, signature_help: #Protocol.Types.SignatureHelp.ClientCapabilities<[context_support: true, dynamic_registration: true, signature_information: #Protocol.Types.SignatureHelp.ClientCapabilities.SignatureInformation<[active_parameter_support: true, documentation_format: [:markdown, :plain_text], parameter_information: #Protocol.Types.SignatureHelp.ClientCapabilities.ParameterInformation<[label_offset_support: true]>]>]>, work_done_progress: true}, client_name: "Visual Studio Code", additional_watched_extensions: nil, dialyzer_enabled?: false}, initialized?: true, shutdown_received?: false, in_flight_requests: %{}}}
```

Seems like we get a DidChange before we get a DidOpen...

scottming commented:
> Seems like we get a DidChange before we get a DidOpen, or we have an open_temporary that expired

Yes, The reason this issue arises is that we don't open a document for the `exact` entry, and vscode first sends DidChange, then DidSave, and finally DidClose. It's possible that some `DidChange` events are sent before `DidOpen`. We solved this issue in the `rename_file` PR when converting `Document.Changes`.

Can you try this operation on that branch?

scottming commented:
@scohen However, with such changes involving hundreds of files, vscode sends hundreds of DidSave messages. Do we need to identify this situation and only execute the first `mix compile` while ignoring the others? Currently, this makes my m3 max scream for a few minutes (due to CPU overload).

scohen commented:
> However, with such changes involving hundreds of files, vscode sends hundreds of DidSave messages

Yes, i saw this as well. We should _definitely_ disable the compilers during the rename operation, and resume it when it's done.

scohen commented:
@scottming I think you can use `Build.with_lock` to disable builds during the rename. You might also need to do something similar for formatting (it doesn't presently have a lock)

scottming commented:
@scohen In this 5b111046cdf5ce75e963c5ca98e864489839c995 of the #651 , I made some attempts and found that using a *delayed* approach works. However, I'm assuming that the client will complete all the renaming operations within two seconds. It works on my computer, and there are no hundreds of compilations anymore. After renaming `Ast` to `AST`, everything works well.

But this method might not be particularly elegant. If you have a better way, I'd be glad to hear it.

scottming committed:
Rename file when renaming module (#651)

* Rename the file while renaming the module.

It first determines whether the module meets certain hard constraints, such as having no parent, then performs some conventional checks. If both are satisfied, it proceeds to return the file renaming.

* Maybe insert special folder for PhoenixWeb's module

* Apply some code review changes for using `Document.Changes`

Use `Document.Changes` instead of `CodeMod.Rename.DocumentChanges`

Make `Document.Changes.RenameFile` a struct and use `root_module?`
instead of `has_silibings?` and `has_parent?`

* Use `rename progress marker` instead of suspending build

* Use `file_changed` instead of `file_compile_requested` for reindexing

* `uri_with_operations_counts` by client name

* Apply some code review suggestions for `rename/file`

* Apply suggestions from code review

Co-authored-by: Steve Cohen <scohen@scohen.org>

* Do not need to care about the `DidClose` message

* Apply some code review suggestions

1. Move the `if-else` logic to the `Commands.Rename` module
2. Add `file_opened` message type
3. modify the format error message

* Shouldn't returns `RenameFile` if the file name not changed

* Change `uri_with_operation_counts` to `uri_with_expected_operation`

and special some message type for emacs

---------

Co-authored-by: Steve Cohen <scohen@scohen.org>

Add `WorkDoneProgress` for the rename expected_operation

Apply suggestions from code review

Co-authored-by: Steve Cohen <scohen@scohen.org>

Update apps/remote_control/lib/lexical/remote_control/code_mod/rename/module.ex

Co-authored-by: Steve Cohen <scohen@scohen.org>

Update apps/remote_control/lib/lexical/remote_control/commands/rename.ex

Co-authored-by: Zach Allaun <zach.allaun@gmail.com>

Use record message instead of Protocol structs

Replace the asynchronous subscription method of `dispatch` with synchronous `updates`.

In the previous commit, we tried the method of registering and subscribing to messages via `dispatch`. I found that this method has significant issues, as it causes the message triggering compilation, the modification of rename status, and the judgment of progress to be out of sync, leading to rename failures and crashes.

Apply another code review suggestions

scottming commented on apps/remote_control/lib/lexical/remote_control/commands/rename.ex:
```diff
@@ -23,10 +38,14 @@ defmodule Lexical.RemoteControl.Commands.Rename do
-        %{^uri => ^operation} -> Map.delete(uri_to_operation, uri)
-        _ -> uri_to_operation
+        %{^uri => ^operation} ->
+          on_update_progess.(1, "Renaming")
```
I think maybe `on_update_progess.(1, "")` is enough?

scohen commented on apps/remote_control/lib/lexical/remote_control/commands/rename.ex:
```diff
@@ -0,0 +1,96 @@
+defmodule Lexical.RemoteControl.Commands.Rename do
+  @moduledoc """
+  We are unable to accurately determine the process of renaming,
+  because after the rename, there will be a series of operations such as
+  `DidChange`, `DidSave`, etc., which will trigger expensive actions like compiling the entire project.
+  Therefore, we need this module to make some markings to determine whether it is currently in the process of renaming.
```
```suggestion
  We are unable to accurately determine when the rename process finishes,
  because after the rename, we will receive a series of events like 
  `DidChange`, `DidSave`, etc., which will trigger expensive actions like compiling the entire project.
  Therefore, we need this module to tell us if lexical is currently in the process of renaming.
```

scohen commented on apps/remote_control/lib/lexical/remote_control/commands/rename.ex:
```diff
@@ -0,0 +1,96 @@
+    {:ok, state}
+  end
+
+  @spec set_rename_progress(%{Lexical.uri() => atom()}, tuple()) :: :ok
```
tuple is a pretty weak spec, what's in it?

scohen commented on apps/server/lib/lexical/server/provider/handlers/prepare_rename.ex:
```diff
@@ -0,0 +1,40 @@
+    case Api.prepare_rename(project, analysis, position) do
+      {:ok, cursor_entity, range} ->
+        result =
+          Types.PrepareRenameResult.PrepareRenameResult.new(
```
alias this

scohen commented on apps/server/lib/lexical/server/provider/handlers/rename.ex:
```diff
@@ -0,0 +1,76 @@
+  end
+
+  defp new_text_document_edit(uri, edits) do
+    text_document = TextDocument.OptionalVersioned.Identifier.new(uri: uri, version: 0)
```
what's the significance of 0 here?

scottming replied:
# NOTE: the `0` here is for use in VSCode.
    # Before May 4, 2024, if this `0` number is not set,
    # the rename function in VSCode will not work,
    # while other editors seem not to care about this field.
Added some comments

scohen replied:
see my other comment.

scohen reviewed (changes requested):
The changes are pertaining to protocol leaking into RemoteControl

scottming commented on apps/remote_control/lib/lexical/remote_control/code_mod/rename/file.ex:
```diff
@@ -0,0 +1,154 @@
+    prefer_to_include_insertions? = insertions && insertions in Path.split(suffix)
+    old_path_contains_insertions? = insertions in Path.split(relative_path)
+
+    if not is_nil(insertions) and old_path_contains_insertions? and
```
I personally prefer a strict 'and', so I used `not is_nil(insertions)` here. Of course, I could also change the default value of insertions to false, although the type `false | String.t()` for `insertions` seems a bit odd.

scottming replied:
I personally prefer a strict 'and' in this whole line.

zachallaun replied:
It looks like insertions is `String.t() | false` at the moment, so `if insertions ...` would work. If you meant for it to be `nil`, it needs to be changed on line 130 above.

zachallaun commented on apps/remote_control/lib/lexical/remote_control/code_mod/rename/file.ex:
```diff
@@ -0,0 +1,150 @@
+    #
+    # iex> fetch_conventional_prefix("apps/remote_control/lib/lexical/remote_control/code_mod/rename/file.ex")
+    # {:ok, "apps/remote_control/lib"}
+    result =
+      path
+      |> Path.split()
+      |> Enum.chunk_every(2, 2)
+      |> Enum.reduce([], fn
+        ["apps", app_name], _ ->
+          [app_name, "apps"]
+
+        ["lib", _follow_element], prefix ->
+          ["lib" | prefix]
+
+        ["test", _follow_element], prefix ->
+          ["test" | prefix]
+
+        _remain, prefix ->
+          prefix
+      end)
+
+    case result do
+      [] ->
+        :error
+
+      prefix ->
+        prefix = prefix |> Enum.reverse() |> Path.join()
+        {:ok, prefix}
+    end
```
Is this more complex than it needs to be? There aren't that many cases, so can we just enumerate them?

```elixir
segments =
  case Path.split(path) do
    ["apps", app_name, "lib" | _] -> ["apps", app_name, "lib"]
    ["apps", app_name, "test" | _] -> ["apps", app_name, "test"]
    ["lib" | _] -> ["lib"]
    ["test" | _] -> ["test"]
    _ -> nil
  end

if segments do
  {:ok, Path.join(segments)}
else
  :error
end
```

scottming replied:
yeah, your solution is more clearer

zachallaun reviewed (commented):
Started reading through this and have a few quick comments. I'm planning to delve deeper, though.

zachallaun commented on apps/remote_control/lib/lexical/remote_control/commands/rename.ex:
```diff
@@ -0,0 +1,96 @@
+defmodule Lexical.RemoteControl.Commands.Rename do
+  @moduledoc """
+  We are unable to accurately determine when the rename process finishes,
+  because after the rename, we will receive a series of events like 
+  `DidChange`, `DidSave`, etc., which will trigger expensive actions like compiling the entire project.
+  Therefore, we need this module to tell us if lexical is currently in the process of renaming.
+  """
```
Small stylistic thing, but I think this should be a comment instead of a moduledoc.

```suggestion
  # We are unable to accurately determine when the rename process finishes,
  # because after the rename, we will receive a series of events like 
  # `DidChange`, `DidSave`, etc., which will trigger expensive actions like compiling the entire project.
  # Therefore, we need this module to tell us if lexical is currently in the process of renaming.
  
```

zachallaun commented on apps/remote_control/lib/lexical/remote_control/build.ex:
```diff
@@ -35,6 +36,22 @@ defmodule Lexical.RemoteControl.Build do
     :ok
   end
 
+  def maybe_schedule_compile(triggered_file_uri, message) do
+    if Rename.in_progress?() do
+      Rename.update_progress(triggered_file_uri, message)
+    else
+      GenServer.cast(__MODULE__, {:compile, false})
+    end
+  end
+
+  def maybe_compile_document(%Project{} = project, %Document{} = document, message) do
+    if Rename.in_progress?() do
+      Rename.update_progress(document.uri, message)
+    else
+      compile_document(project, document)
+    end
+  end
```
This sort of thing makes me a little nervous. There will be other commands/actions/etc. in the future that will take a long time and should potentially block compilation and adding additional checks here will get gross pretty fast.

So is the problem basically that, after issuing all these renames, we're going to get a bunch of messages from the client about all the files we just changed? And we don't want to actually compile until the client is done sending us all of those? I'd really like to consider whether there's a better, more general way to handle this sort of thing. For instance, does a similar thing happen when you checkout a different branch? Lexical gets a bunch of notifications about changed files, starts compiling mid-way through, and sometimes stuff ends up in a sort of borked state until you visit those files again?

This makes me think that compilation, perhaps, should be cancelable/restartable, and whenever a change comes in, it cancels any in-progress compilation and restarts it. So we might get a bunch of cancels/restarts, but now we'd have a system that is more resilient to rapid-fire changes to many files.

zachallaun replied:
> It's not just builds though, it's also formatting, code actions and the like

Yep, all the more reason to have some more general way of indicating that things are in a "partial" state. We definitely don't want every one of these places to have to know about every potentially long-running action.

scottming commented on apps/remote_control/lib/lexical/remote_control/code_mod/rename/module.ex:
```diff
@@ -0,0 +1,182 @@
+    end
+  end
+
+  defp range_text(range) do
```
I don't quite understand what you mean by "guarantee."

We obtain the range for calculating the text through `Rename.Prepare.resolve(analysis, position)`.

scohen replied:
Ranges define two positions, start and end. Both positions have a line of text in them. Is it possible that the module you're looking for doesn't occur on one of those two lines? If that's true, this approach won't work in every case.

scottming commented on apps/remote_control/lib/lexical/remote_control/code_mod/rename/module.ex:
```diff
@@ -0,0 +1,182 @@
+    end
+  end
+
+  defp diff(old_range_text, new_name) do
```
Do you mean to rename the parameter names? like `defp diff(old_name, new_name)`?

scohen replied:
no, we have a whole module that handles diffing two strings, and returns a bunch of edits. It seems you've implemented a bunch of that logic here, though on a much smaller scale, though maybe I don't understand what's going on here.

scottming commented on apps/remote_control/lib/lexical/remote_control/code_mod/rename/module.ex:
```diff
@@ -0,0 +1,182 @@
+
+  defp range_text(range) do
+    line(text: text) = range.end.context_line
+    String.slice(text, range.start.character - 1, range.end.character - range.start.character)
```
It might be because the end is `range.end.character - range.start.character`, so this change won't work and will break all the tests.

scohen replied:
that doesn't make sense to me, `range.end.character - range.start.character` is a length, not a limit in an (elixir) range. 
`String.slice(range.start.character - 1, range.end.character - range.start.character)` should be equivalent to `String.slice(range.start.character - 1..range.end.character)`, no?

scottming replied:
If we're going to use range, then the correct code here would be `String.slice(text, (range.start.character - 1)..(range.end.character - 2))`. Honestly, I think this is more difficult to understand than the previous code.

scohen replied:
why, ranges are inclusive right?

scottming commented on apps/remote_control/lib/lexical/remote_control/api.ex:
```diff
@@ -16,6 +16,21 @@ defmodule Lexical.RemoteControl.Api do
   defdelegate schedule_compile(project, force?), to: Build
   defdelegate compile_document(project, document), to: Build
 
+  def maybe_schedule_compile(project, triggered_file_uri, triggered_message) do
+    RemoteControl.call(project, Build, :maybe_schedule_compile, [
+      triggered_file_uri,
+      triggered_message
+    ])
+  end
+
+  def maybe_compile_document(project, document, triggered_message) do
+    RemoteControl.call(project, Build, :maybe_compile_document, [
+      project,
+      document,
+      triggered_message
+    ])
+  end
```
Really? yes, it can, but we have two `if else` statements there. If we delegate, that means we would need to write more `RemoteControl.call` statements, which would make the code less clear.

scottming commented on apps/remote_control/lib/lexical/remote_control/code_mod/rename.ex:
```diff
@@ -0,0 +1,70 @@
+          {:ok, [Document.Changes.t()]} | {:error, term()}
+  def rename(%Analysis{} = analysis, %Position{} = position, new_name, client_name) do
+    with {:ok, {renamable, entity}, range} <- Rename.Prepare.resolve(analysis, position) do
+      rename_module = @rename_mappings[renamable]
```
I think it's fine to use `@rename_mappings[renamable]` here, since `Rename.Prepare.resolve` will let us know whether the renamable is supported or not.

zachallaun replied:
Got it! In that case, can we use `Map.fetch!` to be explicit that we expect a known key and so any error occurs at that point, instead of below where we try to call a function from the module?

scottming replied:
Ok

scottming commented on apps/remote_control/lib/lexical/remote_control/code_mod/rename/file.ex:
```diff
@@ -0,0 +1,154 @@
+
+  defp root_module?(entry, document) do
+    entries =
+      ProcessCache.trans("#{document.uri}-entries", 50, fn ->
```
This is a cache set up before the rename results are returned, independent of the editor, mainly to prevent repeatedly indexing a file. 50 milliseconds should not be a problem.

scohen replied:
This is why i was asking if the rename process always exists, it might make sense to just not time this out and let the process exiting clear the cache.

scottming replied:
Please note that the process for `Command.Rename` is not the same as this process. The events here occur before `Command.Rename`.

scottming commented on apps/remote_control/lib/lexical/remote_control/code_mod/rename/file.ex:
```diff
@@ -0,0 +1,154 @@
+      new_uri = Document.Path.ensure_uri(new_path)
+
+      if document.uri != new_uri do
+        Document.Changes.RenameFile.new(document.uri, new_uri)
```
Did you mean: `alias  Document.Changes.RenameFile`

scohen replied:
more like `alias Document.Changes`

scohen commented on apps/remote_control/lib/lexical/remote_control/code_mod/rename/prepare.ex:
```diff
@@ -0,0 +1,87 @@
+  @renamable_modules [Rename.Module]
+
+  defp do_resolve(%Analysis{} = analysis, %Position{} = position) do
+    result =
+      Enum.find_value(@renamable_modules, fn module ->
+        result = module.resolve(analysis, position)
+
+        if match?({:ok, _, _}, result) do
+          result
+        end
+      end)
+
+    if is_nil(result) do
+      case Entity.resolve(analysis, position) do
+        {:ok, other, _} ->
+          Logger.info("Unsupported entity for renaming: #{inspect(other)}")
+          {:error, :unsupported_entity}
+
+        {:error, reason} ->
+          {:error, reason}
+      end
+    else
+      result
+    end
```
I _think_ this does the same thing: 

```suggestion
   apply_resolve  = fn module -> 
     case module.resolve(analysis, position) do 
        {:ok, _, _} = result -> result 
         _ -> false 
     end
   end
     
     with :error <- Enum.find_value(@renamable_modules, apply_resolve),
          {:ok, _, _} <- Entity.resolve(analysis, position) do 
          Logger.info("Unsupported entity for renaming: #{inspect(other)}")
          {:error, :unsupported_entity}
     end     
```

scohen commented on apps/remote_control/lib/lexical/remote_control/commands/reindex.ex:
```diff
@@ -64,6 +64,21 @@ defmodule Lexical.RemoteControl.Commands.Reindex do
       end
     end
+
+    defp ensure_document_open(uri) do
```
```suggestion
    defp ensure_open(uri) do
```

scohen commented on apps/remote_control/lib/lexical/remote_control/commands/rename.ex:
```diff
@@ -0,0 +1,114 @@
+  # Instead, it should call this function to synchronously update the status,
+  # thus preventing failures due to latency issues.
+  def update_progress(message) do
+    GenServer.cast(__MODULE__, {:update_progress, message})
```
If the comment is true, shouldn't this be a `GenServer.call` rather than a cast?

scottming replied:
Using cast here is fine; we just need to ensure that the state changes in the `Build` and `Rename` processes occur in sequence. I mean that the `file_changed` message should arrive at `Rename` before `Build`.

scohen commented:
This should pause until we get the proxy done.

scottming committed:
Try integrate proxy instead of multiple `maybe_xxx` method

scohen commented on apps/remote_control/lib/lexical/remote_control/commands/reindex.ex:
```diff
@@ -64,6 +64,21 @@ defmodule Lexical.RemoteControl.Commands.Reindex do
+        {:error, :not_open} ->
+          # Sometimes, some operations are received long after the reindex is triggered,
+          # such as new files after a batch rename
+          with {:ok, document} <- Document.Store.open_temporary(uri) do
+            analysis = Ast.analyze(document)
+            {:ok, document, analysis}
+          end
```
```suggestion
          with {:ok, _} <- Document.Store.open_temporary(uri),
               {:ok, document, analysis} <- Document.Store.fetch(uri, :analysis) do
            {:ok, document, analysis}
          end
```

scohen replied:
though maybe the whole thing should be 
```elixir
with {:error, :not_open} <- Document.Store.fetch(uri, :analysis),
     _ <- Document.Store.open_temporary(uri) do
  Document.Store.fetch(uri, :analysis)
end
```

scohen commented on apps/remote_control/lib/lexical/remote_control/commands/rename.ex:
```diff
@@ -0,0 +1,131 @@
+          state.on_update_progess
+        )
+
+      if new_uri_with_expected_operation == %{} do
```
```suggestion
      if map_size(new_uri_with_expected_operation) == 0 do
```

or 

```suggestion
      if Enum.empty(new_uri_with_expected_operation) do
```

scohen commented on apps/remote_control/lib/lexical/remote_control/commands/rename.ex:
```diff
@@ -0,0 +1,131 @@
+  defmodule State do
+    defstruct uri_with_expected_operation: %{}, on_update_progess: nil, on_complete: nil
+
+    def new(uri_with_expected_operation, progress_functions) do
```
each progress function should be a separate parameter.
a spec here would be nice.

scohen commented on apps/remote_control/lib/lexical/remote_control/commands/rename.ex:
```diff
@@ -0,0 +1,131 @@
+  import Messages
+
+  defmodule State do
+    defstruct uri_with_expected_operation: %{}, on_update_progess: nil, on_complete: nil
```
typically, maps are named like `uri_to_expected_operation`

scohen commented on apps/remote_control/lib/lexical/remote_control/commands/rename.ex:
```diff
@@ -0,0 +1,131 @@
+  alias Lexical.RemoteControl.Api.Proxy
+  use GenServer
+
+  def child_spec(uri_with_expected_operation, progress_functions) do
```
each progress function should be its own parameter.

scohen commented on apps/remote_control/lib/lexical/remote_control/commands/rename.ex:
```diff
@@ -0,0 +1,131 @@
+
+  @impl true
+  def handle_continue(:start_buffering, state) do
+    Proxy.start_buffering(self())
```
Passing in self() here is unnecessary.

scohen commented on apps/remote_control/lib/lexical/remote_control/commands/rename.ex:
```diff
@@ -0,0 +1,131 @@
+  @impl true
+  def handle_continue(:start_buffering, state) do
+    Proxy.start_buffering(self())
+    Logger.info("Rename process started, and started buffering.")
```
remove.

scohen commented on apps/remote_control/lib/lexical/remote_control/commands/rename.ex:
```diff
@@ -0,0 +1,131 @@
+    end
+  end
+
+  def in_progress? do
```
i think this function isn't necessary. Either a rename is in progress, or it's not. If it is, you can return an error, and if it's not, a new rename can be started. It's safer to do as much as you can inside the genserver.

scohen replied:
(in other words, this function is extremely susceptible to races)

scohen commented on apps/remote_control/lib/lexical/remote_control/commands/rename_supervisor.ex:
```diff
@@ -0,0 +1,29 @@
+    DynamicSupervisor.start_link(name: __MODULE__, strategy: :one_for_one)
+  end
+
+  def start_renaming(uri_with_expected_operation, progress_notification_functions) do
```
same comment, the rename functions should be separate parameters.

scohen commented on apps/remote_control/lib/lexical/remote_control/dispatch/handlers/indexing.ex:
```diff
@@ -8,9 +8,20 @@ defmodule Lexical.RemoteControl.Dispatch.Handlers.Indexing do
+  use Dispatch.Handler, [filesystem_event(), file_changed(), file_opened()]
 
-  def on_event(file_compile_requested(uri: uri), state) do
+  def on_event(file_changed(uri: uri, open?: true), state) do
```
why do we care if the file is open to index it?

scottming replied:
Okay, we can reindex all the renamed files once the renaming is complete

scohen commented on apps/remote_control/lib/lexical/remote_control/dispatch/handlers/indexing.ex:
```diff
@@ -8,9 +8,20 @@ defmodule Lexical.RemoteControl.Dispatch.Handlers.Indexing do
+  end
+
+  def on_event(file_opened(uri: uri), state) do
+    # When renaming occurs, new files are only *opened* and *closed*,
+    # so we need to index the new files.
```
what about when it's not occurring? 
I think a better idea is to wait until the rename is done, then reindex all the files that changed.

scohen commented on apps/server/lib/lexical/server/provider/handlers/prepare_rename.ex:
```diff
@@ -0,0 +1,40 @@
+        {:reply, Responses.PrepareRename.new(id, result)}
+
+      _ ->
+        {:reply, Responses.PrepareRename.new(id, nil)}
```
if this is the error case, should we respond with a message like the docs say to do?

scottming replied:
If I understand the document correctly, then for modules that are not at the defined location, we should return null, while for other entities, we should return an error message.

scohen replied:
Yes

scottming replied:
Updated. Please take a look at the latest commit.

![CleanShot 2024-06-20 at 14 02 26@2x](https://github.com/lexical-lsp/lexical/assets/12830256/3de333b7-68b8-42c5-b48f-ea99a91a2cea)

scohen commented on apps/server/lib/lexical/server/provider/handlers/rename.ex:
```diff
@@ -0,0 +1,80 @@
+    case Document.Store.fetch(request.document.uri, :analysis) do
+      {:ok, _document, %Ast.Analysis{valid?: true} = analysis} ->
+        rename(
+          env.project,
+          analysis,
+          request.position,
+          request.new_name,
+          request.id,
+          env.client_name
```
Why not pass the env and request?
i.e. 
```elixir
def rename(%Rename{} = request, %Env{} = env) do
...
```

scohen commented on apps/server/lib/lexical/server/provider/handlers/rename.ex:
```diff
@@ -0,0 +1,80 @@
+
+  defp new_text_document_edit(uri, edits) do
+    # NOTE: the `0` here is for use in VSCode.
+    # Before May 4, 2024, if this `0` number is not set,
+    # the rename function in VSCode will not work,
```
you sure that the reason is that you're not setting the correct version of opened documents?

scottming replied:
You are right, simply using `document.version` here will suffice, fixed

scohen commented on apps/server/lib/lexical/server/provider/handlers/rename.ex:
```diff
@@ -0,0 +1,80 @@
+        {:reply, nil}
+
+      {:ok, results} ->
+        text_document_edits =
+          Enum.map(results, fn %Document.Changes{edits: edits, document: document} ->
+            new_text_document_edit(document.uri, edits)
+          end)
+
+        rename_files =
+          results
+          |> Stream.map(& &1.rename_file)
+          |> Stream.reject(&(&1 == nil))
+          |> Enum.map(&new_rename_file/1)
```
```suggestion
        rename_files = for %{rename_file: rename_file} <- results,
             not is_nil(rename_file) do
             rename_file
        end
```

scohen replied:
Can this just be:

```suggestion
       document_changes = Enum.flat_map(results, fn 
        %Document.Changes{rename_file: %Document.Changes.RenameFile{}} = changes ->
           [new_text_document_edit(changes), new_rename_file(changes.rename_file)]
         %Document.Changes{} = changes -> 
           [new_text_document_changes(changes)]
         end)
         
        workspace_edit = Workspace.Edit.new(document_changes: document_changes)        
```

scohen commented on apps/server/lib/lexical/server/provider/handlers/rename.ex:
```diff
@@ -0,0 +1,80 @@
+        {:reply, Responses.Rename.new(id, workspace_edit)}
+
+      {:error, {:unsupported_entity, entity}} ->
+        Logger.info("Unrenameable entity: #{inspect(entity)}")
```
```suggestion
        Logger.info("Cannot rename entity: #{inspect(entity)}")
```

scohen commented on apps/server/lib/lexical/server/provider/handlers/rename.ex:
```diff
@@ -0,0 +1,80 @@
+    end
+  end
+
+  defp new_text_document_edit(uri, edits) do
```
```suggestion
  defp new_text_document_edit(%Document.Changes{} = changes) do
```

scottming commented on apps/remote_control/lib/lexical/remote_control/code_mod/rename/module.ex:
```diff
@@ -0,0 +1,193 @@
+  alias Lexical.RemoteControl.CodeIntelligence.Entity
+  alias Lexical.RemoteControl.CodeMod.Rename
+  alias Lexical.RemoteControl.Search.Store
+  require Logger
+
+  import Line
```
Shouldn't `require` come before `import`? I was just following a suggestion from younger Steven: https://github.com/elixir-lsp/elixir-ls/pull/813/files/e468940cf4c47bc0836f08144e357a94babc1320#r1130348552

scohen replied:
alright, ignore.

scottming commented on apps/remote_control/lib/lexical/remote_control/code_intelligence/entity.ex:
```diff
@@ -52,6 +52,22 @@ defmodule Lexical.RemoteControl.CodeIntelligence.Entity do
     )
   end
 
+  @spec phoenix_controller_module?(module()) :: boolean()
+  def phoenix_controller_module?(module) do
+    function_exists?(module, :call, 2) and function_exists?(module, :action, 2)
+  end
+
+  @spec phoenix_liveview_module?(module()) :: boolean()
+  def phoenix_liveview_module?(module) do
+    function_exists?(module, :mount, 3) and function_exists?(module, :render, 1)
+  end
+
+  @spec phoenix_component_module?(module()) :: boolean()
+  def phoenix_component_module?(module) do
+    function_exists?(module, :__components__, 0) and
+      function_exists?(module, :__phoenix_component_verify__, 1)
+  end
+
```
related to the `Rename file`

scohen replied:
how so?

scottming replied:
If a file is a component, then when renaming it, we should not change it to `Components.RenamedModule`, nor should we move it out of the components folder

scottming committed:
Apply suggestions from code review

Co-authored-by: Steve Cohen <scohen@scohen.org>

scottming commented on apps/remote_control/lib/lexical/remote_control/code_mod/rename/prepare.ex:
```diff
@@ -0,0 +1,78 @@
+  defp cursor_at_declaration?(module, rename_range) do
+    case Store.exact(module, type: :module, subtype: :definition) do
+      {:ok, [definition]} ->
+        rename_range == definition.range
```
This method is reliable because it compares the range of the module after 'surround' with the defined range of 'exact'. However, do you have any better suggestions for the comparison?

scohen replied:
struct comparison does a deep compare, there are fields in the `Position` struct that shouldn't have a bearing on them being equal. I think you want to compare ranges based on their positions' lines and characters.

scottming replied:
@scohen So should we implement an equal? function for range or position to do this?

zachallaun replied:
@scottming #767 implements this exact function but is currently dependent on #768. It's only dependent, however, because #767 refactors `Range.contains?/2` to be in terms of `Position.compare/2`, which causes tests to fail without #768.

I can very easily move things around such that #767 can be merged in first without #768, then you can rebase and use that function here.

zachallaun replied:
#767 has been updated so that it can be merged independently if someone is able to review it.

scottming replied:
@zachallaun thanks

scottming commented on apps/remote_control/lib/lexical/remote_control/code_mod/rename/prepare.ex:
```diff
@@ -0,0 +1,78 @@
+    # not just the last local module, like: `defmodule |Foo.Bar do` also works.
+    {:ok, %{end: {_end_line, end_character}}} = Ast.surround_context(analysis, position)
+    end_position = %{position | character: end_character - 1}
+    {:ok, {:module, module}, range} = do_resolve(analysis, end_position)
```
This is the resolve for different locations, so multiple occurrences here are reasonable.

scottming committed:
Reindex all renamed files after renaming done

scottming committed:
Use `document.version` instead of `0`

scottming commented on apps/remote_control/lib/lexical/remote_control/code_mod/rename/prepare.ex:
```diff
@@ -0,0 +1,78 @@
+
+  @spec resolve(Analysis.t(), Position.t()) ::
+          {:ok, {atom(), atom()} | {atom(), tuple()}, Range.t()} | {:error, term()}
+  def resolve(%Analysis{} = analysis, %Position{} = position) do
+    case do_resolve(analysis, position) do
+      {:ok, {:module, _module}, _range} ->
+        {module, range} = surround_the_whole_module(analysis, position)
+
+        if cursor_at_declaration?(module, range) do
+          {:ok, {:module, module}, range}
+        else
+          {:error, {:unsupported_location, :module}}
+        end
+
+      other ->
+        other
+    end
```
Can you explain this comment in detail? I don't quite understand it.

scottming replied:
@scohen Updated. [ebf68fd](https://github.com/lexical-lsp/lexical/pull/636/commits/ebf68fdc66d526ef9252d198202cf79642642432)
This indeed simplifies the content of the prepare module, but the downside is that we will call Entity.resolve twice. However, I think this is a minor issue.

scottming committed:
Enhancement the `PrepareRename` error message

When the entity at the cursor supports renaming but is at the wrong
location, return nil. When the entity is of an unsupported type, return
an appropriate error.

scottming committed:
Use `recognizes?` and `prepare` pattern

scohen commented:
I got a chance to test this out extensively today. 

1. I had to switch to vscode, emacs was just too unreliable. Pity. I think the problems belong to emacs and not lexical. Vscode is fast and accurate.
1. It seems to rename modules that are aliased incorrectly, going through and first replacing the alias with the new module name, but then going through each alias and expanding it.  so if I `alias Foo.Bar.Utils` and later in the module call `Utils.blah()` and rename `Foo.Bar.Utils` to `Quux.Utils` the alias is correctly expanded to `alias Quux.Utils` but the call is also expanded to `Quux.Utils.blah()`. 
2. I was renaming modules in `test/support`. Modules had no previous structure, and I was adding `Company.Testing.OldName`. It placed the results in `test/company/OldName.ex`, i would have preferred to keep them in `test/support`. I think this is due to the path detection not recognizing `test/support` as a root. 
3. To ensure the indexer was up to date, i had to reindex after each rename. It would be better to wait for the rename to end, then reindex the files that have changed.
4. It doesn't seem to handle renaming of modules in multi-module aliases. For example given `alias Company.{A, B}` and you alias `A to Company.Something.A` the alias won't be affected. It might be because the alias wash `Company.{A, B}` and the rename was to `CompanyWeb.Testing.A`
5. **Tough**: Renaming a module often will mean it needs to be reordered in its context. It would be cool if it's in an alias / import / require / use group if it could be alphabetized.

scottming commented:
> It seems to rename modules that are aliased incorrectly, going through and first replacing the alias with the new module name, but then going through each alias and expanding it. so if I alias Foo.Bar.Utils and later in the module call Utils.blah() and rename Foo.Bar.Utils to Quux.Utils the alias is correctly expanded to alias Quux.Utils but the call is also expanded to Quux.Utils.blah().

This issue you mentioned do exist and are indeed difficult to handle.

> To ensure the indexer was up to date, i had to reindex after each rename. It would be better to wait for the rename to end, then reindex the files that have changed.

This is what we're currently doing; you don't need to reindex every time before renaming."

scohen commented:
> This issue you mentioned do exist and are indeed difficult to handle.

Can you elaborate? 

> This is what we're currently doing; you don't need to reindex every time before renaming.

Then it doesn't seem to be work, the index was most definitely out of date.

scottming committed:
Rebase main and rename modify the `%Env{}` refs to `%Configuratoin{}`

scottming committed:
Add progress tracking for reindexing as well

scottming commented:
> Can you elaborate? After the rename, we should have knowledge of what the aliases at a given line are. You should look for the module you want to replace in the aliases at the line, and if it's present, use the alias, if not, use the full module.

@scohen The reason I say this issue is difficult to handle is because renaming encompasses many scenarios, such as:

1. Renaming the last module: `Foo.Bar` -> `Foo.Renamed`. This is also the one I use most frequently.
2. Adding some modules in the middle: `Foo.Bar` -> `Foo.Baz.Bar`. This is the situation you mentioned that is prone to errors, mainly regarding references in aliased places.
3. Removing some modules in the middle: `Foo.Baz.Bar` -> `Foo.Bar`.
4. Replacing most of the modules in the suffix: `Foo.Baz.Bar` -> `Foo.Renamed`.
5. Other possible scenarios.

For the second scenario, if there is an `alias Foo.Bar` in a file, then indeed, references to `Bar` in other places will become `Baz.Bar`. This is unnecessary, but the problem is that I don't know of a good way to accurately distinguish the second renaming method from the others. Given that such errors will be discovered after compilation, if there isn't a good way to identify them, I don't want to add more complexity.

scottming committed:
Use a brand new diff strategy and apply some special handling for renaming non-alias reference modules.

This can fix some errors that occur when renaming without changing the local name.

scottming committed:
Fix the work progress for neovim

scottming committed:
Try fix the indexing after renaming for emacs
